docs/Contributing/Adding-new-endpoints.md

```diff
@@ -0,0 +1,255 @@
+        return nil, err
+    }
+
+    return svc.ds.CountAllHosts(ctx)
+}
+```
+
+As you can see the only thing this method adds to the table is the authorization. We assume that if a user can list 
+hosts then they could list all hosts and count them, so they can have Fleet do that math for them.
+
+Just like with the `Datastore`, we need to add this method in the `Service` interface as well because otherwise we won't 
+be able to call this from the endpoint function itself:
+
+```go
+type Service interface {
+	// rest of the interface here
+
+	CountAllHosts(ctx context.Context) (int, error)
+}
+```
+
+Now we are ready to work on the endpoint itself.
+
+### Step 3: Endpoint
+
+We're going to be working in the same [hosts.go file from before](https://github.com/fleetdm/fleet/blob/main/server/service/hosts.go):
+
+```go
+/////////////////////////////////////////////////////////////////////////////////
+// Count total amount of hosts
+/////////////////////////////////////////////////////////////////////////////////
+
+type countAllHostsRequest struct {}
+
+type countAllHostsResponse struct {
+    Err error `json:"error,omitempty"`
+    Count int `json:"count"`
+}
+
+func (r countAllHostsResponse) error() error { return r.Err }
+
+func countAllHostsEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (interface{}, error) {
+    req := request.(*countAllHostsRequest)
+    count, err := svc.CountAllHosts(ctx)
+    if err != nil {
+        return countAllHostsResponse{Err: err}, nil
+    }
+    return countAllHostsResponse{Count: count}, nil
+}
+
+func (svc *Service) CountAllHosts(ctx context.Context) (int, error) {
+	// ...
+}
+```
+
+We added 4 things above:
+
+1. The struct that represents details about the request that we might receive. If the request could have query 
+parameters, or a JSON body, etc, this is where we would define it.
+2. The struct for the response. This struct has to implement the `errorer` [interface](https://github.com/fleetdm/fleet/blob/main/server/service/transport_error.go#L21).
+3. The implementation of the only method in the `errorer` interface.
+4. The endpoint function handler itself.
+
+Now it's time to expose this to be used.
+
+### Step 4: Exposing the new API
+
+The whole API can be seen in the [handler.go file](https://github.com/fleetdm/fleet/blob/main/server/service/handler.go).
+Most there is to know about what API is being expose lives in this file and particularly within the 
+`attachFleetAPIRoutes`.
+
+Since this endpoint is a user authenticated endpoint, we'll add it at the end of the `ue` (user authenticated) endpoints 
+of the function mentioned:
+
+```go
+func attachFleetAPIRoutes(r *mux.Router, svc fleet.Service, config config.FleetConfig,
+	logger kitlog.Logger, limitStore throttled.GCRAStore, opts []kithttp.ServerOption,
+	extra extraHandlerOpts,
+) {
+	// ...
+
+	ue.GET("/api/_version_/fleet/hosts/count_all", countAllHostsEndpoint, countAllHostsRequest)
```

Comment:
We should probably say a word about the `_version_` placeholder in the path here, and point to the versioning document for more details.